### ISSUE TITLE
New collection API requires andThen wrappers

### ISSUE THREAD

**Issue by backspace** [san650/ember-cli-page-object#377]
Hello again! I love the new collection API, so much cleaner. In the course of updating the `travis-web` page objects, though, I found that I needed to add a bunch of `andThen`s that were previously unnecessary, whenever we were using a page object to perform an action.

For instance, in a test that cancels a build, with the old API, we have [this](https://github.com/travis-ci/travis-web/blob/293e373b85b162df56b33895c85398b8c05aee8a/tests/acceptance/repo/build-list-routes-test.js#L274):

```
  page.builds(0).cancelButton.click();
```

I mostly used regexes to update the uses of collections across the codebase, which changed it to this:

```
  page.builds[0].cancelButton.click();
```

However, the test was failing because that element wasn’t found! In all the places where we used page objects within a collection, I had to add an `andThen` wrapper; as you can see in [the commit](https://github.com/travis-ci/travis-web/pull/1516/commits/b54419030c7f18ccc151a956887360283fe56a06), there were quite a few instances.

I isolated the above example in [this commit](https://github.com/travis-ci/travis-web/pull/1516/commits/f34b4e3e26472f3b20b1d0f5d8c6596c60717199) to show [the failure on Travis](https://travis-ci.org/travis-ci/travis-web/builds/338112512#L571-L574). I also [tried](https://github.com/travis-ci/travis-web/pull/1516/commits/55f648ac1c5ef49326d36e4829a99c8749c2ec65) the non-proxy version, but the error persisted.

Is this intended behaviour? It seems like a pretty dramatic change.

**Comment by san650** [comment]
This is really interesting, let me take a look to understand what's happening and come back to you.

I liked the old approach without the extra `andThen`, let me see how easy is to bring that back.

**Comment by ro0gr** [comment]
@backspace great report! Thanks.

I think the issue is that we changed the way we handle case when the collection item is not found. 

In the legacy collection we always return a page object instance despite the equal DOM node exists or not. That allows us to reach `click()` async helper:

```
page.builds(0) // returns a PO
  .cancelButton // returns a PO
  .click(); // an async property. It's not executed until all the async is settled
```

With the new collection we started to throw in case if the collection item is not found which prevents us to reach async `click()`. I think we should preserve the existing behavior.

Related discussion: https://github.com/san650/ember-cli-page-object/pull/347#discussion_r162395994

**Comment by backspace** [comment]
It’s strange though because before these changes, the elements _did_ exist. For the example I linked to, the cancel button was definitely being clicked because the flash message that shows after cancellation was properly displayed.

But I regularly have confusion about what I can expect to be available outside `andThen` so maybe this is another instance of that 😆

Thanks for the thread link; I skimmed a bit of the discussion around this, but it was enormous! I appreciate all the thought and work that’s gone into this.
